iroh-net/src/hp/magicsock/endpoint.rs

```diff
@@ -125,39 +130,113 @@ impl Endpoint {
 
     /// Returns the address(es) that should be used for sending the next packet.
     /// Zero, one, or both of UDP address and DERP addr may be non-zero.
-    fn addr_for_send(&self, now: &Instant) -> (Option<SocketAddr>, Option<SocketAddr>) {
-        let udp_addr = self.best_addr.as_ref().map(|a| a.addr);
-        let mut derp_addr = None;
-        if udp_addr.is_none() || !self.is_best_addr_valid(*now) {
-            debug!(
-                "no good udp addr {:?} {:?} - {:?}",
-                now, udp_addr, self.trust_best_addr_until
-            );
-            // We had a best_addr but it expired so send both to it and DERP.
-            derp_addr = self.derp_addr;
+    fn addr_for_send(&mut self, now: &Instant) -> (Option<SocketAddr>, Option<SocketAddr>, bool) {
+        match self.best_addr {
+            Some(ref best_addr) => {
+                if !self.is_best_addr_valid(*now) {
+                    // We had a best_addr but it expired so send both to it and DERP.
+                    debug!(
+                        "best addr is outdated {:?} {:?} - {:?}",
+                        now, best_addr, self.trust_best_addr_until
+                    );
+
+                    (Some(best_addr.addr), self.derp_addr, true)
+                } else {
+                    // Address is current and can be used
+                    (Some(best_addr.addr), None, false)
+                }
+            }
+            None => {
+                let (addr, should_ping) = self.get_candidate_udp_addr(now);
+
+                // provide backup derp addr if no known latency or no addr
+                let derp_addr = if should_ping || addr.is_none() {
+                    self.derp_addr
+                } else {
+                    None
+                };
+
+                (addr, derp_addr, should_ping)
+            }
+        }
+    }
+
+    /// Determines a potential best addr for this endpoint. And if the endpoint needs a ping.
+    fn get_candidate_udp_addr(&mut self, now: &Instant) -> (Option<SocketAddr>, bool) {
+        let mut udp_addr = None;
+        let mut lowest_latency = Duration::from_secs(60 * 60);
+        for (ipp, state) in self.endpoint_state.iter() {
+            if let Some(latency) = state.latency() {
+                // Lower latency, or when equal, prever IPv6.
+                if latency < lowest_latency || (latency == lowest_latency && ipp.is_ipv6()) {
+                    lowest_latency = latency;
+                    udp_addr.replace(*ipp);
+                }
+            }
+        }
+
+        // If we found a candidate, set to best addr
+        if let Some(addr) = udp_addr {
+            self.best_addr = Some(AddrLatency {
+                addr,
+                latency: Some(lowest_latency),
+            });
+            self.trust_best_addr_until
+                .replace(*now + Duration::from_secs(60 * 60));
```

Comment:
maybe trust from `last_ping` time, rather than now? & if that time exceeds an hour it is no longer a `best_addr` candidate?

Reply:
fixed